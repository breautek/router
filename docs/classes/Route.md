[@breautek/router - v3.0.1](../README.md) / [Exports](../modules.md) / Route

# Class: Route<TComponentProps, TRouteProps, TRouteState\>

This class represents a route that renders a [View](View.md) component

## Type parameters

| Name | Type |
| :------ | :------ |
| `TComponentProps` | extends [`IComponentProps`](../interfaces/IComponentProps.md) = [`IComponentProps`](../interfaces/IComponentProps.md) |
| `TRouteProps` | extends [`IRouteProps`](../interfaces/IRouteProps.md)<`TComponentProps`\> = [`IRouteProps`](../interfaces/IRouteProps.md)<`TComponentProps`\> |
| `TRouteState` | extends [`IRouteState`](../interfaces/IRouteState.md) = [`IRouteState`](../interfaces/IRouteState.md) |

## Hierarchy

- `Component`<`TRouteProps`, `TRouteState`\>

  ↳ **`Route`**

## Table of contents

### Constructors

- [constructor](Route.md#constructor)

### Properties

- [context](Route.md#context)
- [props](Route.md#props)
- [refs](Route.md#refs)
- [state](Route.md#state)
- [contextType](Route.md#contexttype)

### Methods

- [UNSAFE\_componentWillMount](Route.md#unsafe_componentwillmount)
- [UNSAFE\_componentWillReceiveProps](Route.md#unsafe_componentwillreceiveprops)
- [UNSAFE\_componentWillUpdate](Route.md#unsafe_componentwillupdate)
- [componentDidCatch](Route.md#componentdidcatch)
- [componentDidMount](Route.md#componentdidmount)
- [componentDidUpdate](Route.md#componentdidupdate)
- [componentWillMount](Route.md#componentwillmount)
- [componentWillReceiveProps](Route.md#componentwillreceiveprops)
- [componentWillUnmount](Route.md#componentwillunmount)
- [componentWillUpdate](Route.md#componentwillupdate)
- [forceUpdate](Route.md#forceupdate)
- [getSnapshotBeforeUpdate](Route.md#getsnapshotbeforeupdate)
- [getView](Route.md#getview)
- [render](Route.md#render)
- [setState](Route.md#setstate)
- [shouldComponentUpdate](Route.md#shouldcomponentupdate)

## Constructors

### constructor

• **new Route**<`TComponentProps`, `TRouteProps`, `TRouteState`\>(`props`)

#### Type parameters

| Name | Type |
| :------ | :------ |
| `TComponentProps` | extends [`IComponentProps`](../interfaces/IComponentProps.md) = [`IComponentProps`](../interfaces/IComponentProps.md) |
| `TRouteProps` | extends [`IRouteProps`](../interfaces/IRouteProps.md)<`TComponentProps`, `TRouteProps`\> = [`IRouteProps`](../interfaces/IRouteProps.md)<`TComponentProps`\> |
| `TRouteState` | extends [`IRouteState`](../interfaces/IRouteState.md) = [`IRouteState`](../interfaces/IRouteState.md) |

#### Parameters

| Name | Type |
| :------ | :------ |
| `props` | `TRouteProps` |

#### Overrides

React.Component&lt;TRouteProps, TRouteState\&gt;.constructor

#### Defined in

[src/Route.tsx:41](https://github.com/breautek/router/blob/c2e6307/src/Route.tsx#L41)

## Properties

### context

• **context**: `any`

If using the new style context, re-declare this in your class to be the
`React.ContextType` of your `static contextType`.
Should be used with type annotation or static contextType.

```ts
static contextType = MyContext
// For TS pre-3.7:
context!: React.ContextType<typeof MyContext>
// For TS 3.7 and above:
declare context: React.ContextType<typeof MyContext>
```

**`see`** https://reactjs.org/docs/context.html

#### Inherited from

React.Component.context

#### Defined in

node_modules/@types/react/index.d.ts:480

___

### props

• `Readonly` **props**: `Readonly`<`TRouteProps`\> & `Readonly`<{ `children?`: `ReactNode`  }\>

#### Inherited from

React.Component.props

#### Defined in

node_modules/@types/react/index.d.ts:505

___

### refs

• **refs**: `Object`

**`deprecated`**
https://reactjs.org/docs/refs-and-the-dom.html#legacy-api-string-refs

#### Index signature

▪ [key: `string`]: `ReactInstance`

#### Inherited from

React.Component.refs

#### Defined in

node_modules/@types/react/index.d.ts:511

___

### state

• **state**: `Readonly`<`TRouteState`\>

#### Inherited from

React.Component.state

#### Defined in

node_modules/@types/react/index.d.ts:506

___

### contextType

▪ `Static` `Optional` **contextType**: `Context`<`any`\>

If set, `this.context` will be set at runtime to the current value of the given Context.

Usage:

```ts
type MyContext = number
const Ctx = React.createContext<MyContext>(0)

class Foo extends React.Component {
  static contextType = Ctx
  context!: React.ContextType<typeof Ctx>
  render () {
    return <>My context's value: {this.context}</>;
  }
}
```

**`see`** https://reactjs.org/docs/context.html#classcontexttype

#### Inherited from

React.Component.contextType

#### Defined in

node_modules/@types/react/index.d.ts:462

## Methods

### UNSAFE\_componentWillMount

▸ `Optional` **UNSAFE_componentWillMount**(): `void`

Called immediately before mounting occurs, and before `Component#render`.
Avoid introducing any side-effects or subscriptions in this method.

This method will not stop working in React 17.

Note: the presence of getSnapshotBeforeUpdate or getDerivedStateFromProps
prevents this from being invoked.

**`deprecated`** 16.3, use componentDidMount or the constructor instead

**`see`** https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#initializing-state

**`see`** https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#gradual-migration-path

#### Returns

`void`

#### Inherited from

React.Component.UNSAFE\_componentWillMount

#### Defined in

node_modules/@types/react/index.d.ts:716

___

### UNSAFE\_componentWillReceiveProps

▸ `Optional` **UNSAFE_componentWillReceiveProps**(`nextProps`, `nextContext`): `void`

Called when the component may be receiving new props.
React may call this even if props have not changed, so be sure to compare new and existing
props if you only want to handle changes.

Calling `Component#setState` generally does not trigger this method.

This method will not stop working in React 17.

Note: the presence of getSnapshotBeforeUpdate or getDerivedStateFromProps
prevents this from being invoked.

**`deprecated`** 16.3, use static getDerivedStateFromProps instead

**`see`** https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#updating-state-based-on-props

**`see`** https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#gradual-migration-path

#### Parameters

| Name | Type |
| :------ | :------ |
| `nextProps` | `Readonly`<`TRouteProps`\> |
| `nextContext` | `any` |

#### Returns

`void`

#### Inherited from

React.Component.UNSAFE\_componentWillReceiveProps

#### Defined in

node_modules/@types/react/index.d.ts:748

___

### UNSAFE\_componentWillUpdate

▸ `Optional` **UNSAFE_componentWillUpdate**(`nextProps`, `nextState`, `nextContext`): `void`

Called immediately before rendering when new props or state is received. Not called for the initial render.

Note: You cannot call `Component#setState` here.

This method will not stop working in React 17.

Note: the presence of getSnapshotBeforeUpdate or getDerivedStateFromProps
prevents this from being invoked.

**`deprecated`** 16.3, use getSnapshotBeforeUpdate instead

**`see`** https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#reading-dom-properties-before-an-update

**`see`** https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#gradual-migration-path

#### Parameters

| Name | Type |
| :------ | :------ |
| `nextProps` | `Readonly`<`TRouteProps`\> |
| `nextState` | `Readonly`<`TRouteState`\> |
| `nextContext` | `any` |

#### Returns

`void`

#### Inherited from

React.Component.UNSAFE\_componentWillUpdate

#### Defined in

node_modules/@types/react/index.d.ts:776

___

### componentDidCatch

▸ `Optional` **componentDidCatch**(`error`, `errorInfo`): `void`

Catches exceptions generated in descendant components. Unhandled exceptions will cause
the entire component tree to unmount.

#### Parameters

| Name | Type |
| :------ | :------ |
| `error` | `Error` |
| `errorInfo` | `ErrorInfo` |

#### Returns

`void`

#### Inherited from

React.Component.componentDidCatch

#### Defined in

node_modules/@types/react/index.d.ts:645

___

### componentDidMount

▸ `Optional` **componentDidMount**(): `void`

Called immediately after a component is mounted. Setting state here will trigger re-rendering.

#### Returns

`void`

#### Inherited from

React.Component.componentDidMount

#### Defined in

node_modules/@types/react/index.d.ts:624

___

### componentDidUpdate

▸ `Optional` **componentDidUpdate**(`prevProps`, `prevState`, `snapshot?`): `void`

Called immediately after updating occurs. Not called for the initial render.

The snapshot is only present if getSnapshotBeforeUpdate is present and returns non-null.

#### Parameters

| Name | Type |
| :------ | :------ |
| `prevProps` | `Readonly`<`TRouteProps`\> |
| `prevState` | `Readonly`<`TRouteState`\> |
| `snapshot?` | `any` |

#### Returns

`void`

#### Inherited from

React.Component.componentDidUpdate

#### Defined in

node_modules/@types/react/index.d.ts:687

___

### componentWillMount

▸ `Optional` **componentWillMount**(): `void`

Called immediately before mounting occurs, and before `Component#render`.
Avoid introducing any side-effects or subscriptions in this method.

Note: the presence of getSnapshotBeforeUpdate or getDerivedStateFromProps
prevents this from being invoked.

**`deprecated`** 16.3, use componentDidMount or the constructor instead; will stop working in React 17

**`see`** https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#initializing-state

**`see`** https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#gradual-migration-path

#### Returns

`void`

#### Inherited from

React.Component.componentWillMount

#### Defined in

node_modules/@types/react/index.d.ts:702

___

### componentWillReceiveProps

▸ `Optional` **componentWillReceiveProps**(`nextProps`, `nextContext`): `void`

Called when the component may be receiving new props.
React may call this even if props have not changed, so be sure to compare new and existing
props if you only want to handle changes.

Calling `Component#setState` generally does not trigger this method.

Note: the presence of getSnapshotBeforeUpdate or getDerivedStateFromProps
prevents this from being invoked.

**`deprecated`** 16.3, use static getDerivedStateFromProps instead; will stop working in React 17

**`see`** https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#updating-state-based-on-props

**`see`** https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#gradual-migration-path

#### Parameters

| Name | Type |
| :------ | :------ |
| `nextProps` | `Readonly`<`TRouteProps`\> |
| `nextContext` | `any` |

#### Returns

`void`

#### Inherited from

React.Component.componentWillReceiveProps

#### Defined in

node_modules/@types/react/index.d.ts:731

___

### componentWillUnmount

▸ `Optional` **componentWillUnmount**(): `void`

Called immediately before a component is destroyed. Perform any necessary cleanup in this method, such as
cancelled network requests, or cleaning up any DOM elements created in `componentDidMount`.

#### Returns

`void`

#### Inherited from

React.Component.componentWillUnmount

#### Defined in

node_modules/@types/react/index.d.ts:640

___

### componentWillUpdate

▸ `Optional` **componentWillUpdate**(`nextProps`, `nextState`, `nextContext`): `void`

Called immediately before rendering when new props or state is received. Not called for the initial render.

Note: You cannot call `Component#setState` here.

Note: the presence of getSnapshotBeforeUpdate or getDerivedStateFromProps
prevents this from being invoked.

**`deprecated`** 16.3, use getSnapshotBeforeUpdate instead; will stop working in React 17

**`see`** https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#reading-dom-properties-before-an-update

**`see`** https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#gradual-migration-path

#### Parameters

| Name | Type |
| :------ | :------ |
| `nextProps` | `Readonly`<`TRouteProps`\> |
| `nextState` | `Readonly`<`TRouteState`\> |
| `nextContext` | `any` |

#### Returns

`void`

#### Inherited from

React.Component.componentWillUpdate

#### Defined in

node_modules/@types/react/index.d.ts:761

___

### forceUpdate

▸ **forceUpdate**(`callback?`): `void`

#### Parameters

| Name | Type |
| :------ | :------ |
| `callback?` | () => `void` |

#### Returns

`void`

#### Inherited from

React.Component.forceUpdate

#### Defined in

node_modules/@types/react/index.d.ts:497

___

### getSnapshotBeforeUpdate

▸ `Optional` **getSnapshotBeforeUpdate**(`prevProps`, `prevState`): `any`

Runs before React applies the result of `render` to the document, and
returns an object to be given to componentDidUpdate. Useful for saving
things such as scroll position before `render` causes changes to it.

Note: the presence of getSnapshotBeforeUpdate prevents any of the deprecated
lifecycle events from running.

#### Parameters

| Name | Type |
| :------ | :------ |
| `prevProps` | `Readonly`<`TRouteProps`\> |
| `prevState` | `Readonly`<`TRouteState`\> |

#### Returns

`any`

#### Inherited from

React.Component.getSnapshotBeforeUpdate

#### Defined in

node_modules/@types/react/index.d.ts:681

___

### getView

▸ **getView**(): [`View`](View.md)<[`IViewProps`](../interfaces/IViewProps.md)\>

#### Returns

[`View`](View.md)<[`IViewProps`](../interfaces/IViewProps.md)\>

#### Defined in

[src/Route.tsx:50](https://github.com/breautek/router/blob/c2e6307/src/Route.tsx#L50)

___

### render

▸ **render**(): `ReactNode`

#### Returns

`ReactNode`

#### Overrides

React.Component.render

#### Defined in

[src/Route.tsx:46](https://github.com/breautek/router/blob/c2e6307/src/Route.tsx#L46)

___

### setState

▸ **setState**<`K`\>(`state`, `callback?`): `void`

#### Type parameters

| Name | Type |
| :------ | :------ |
| `K` | extends `string` \| `number` \| `symbol` |

#### Parameters

| Name | Type |
| :------ | :------ |
| `state` | `TRouteState` \| (`prevState`: `Readonly`<`TRouteState`\>, `props`: `Readonly`<`TRouteProps`\>) => `TRouteState` \| `Pick`<`TRouteState`, `K`\> \| `Pick`<`TRouteState`, `K`\> |
| `callback?` | () => `void` |

#### Returns

`void`

#### Inherited from

React.Component.setState

#### Defined in

node_modules/@types/react/index.d.ts:492

___

### shouldComponentUpdate

▸ `Optional` **shouldComponentUpdate**(`nextProps`, `nextState`, `nextContext`): `boolean`

Called to determine whether the change in props and state should trigger a re-render.

`Component` always returns true.
`PureComponent` implements a shallow comparison on props and state and returns true if any
props or states have changed.

If false is returned, `Component#render`, `componentWillUpdate`
and `componentDidUpdate` will not be called.

#### Parameters

| Name | Type |
| :------ | :------ |
| `nextProps` | `Readonly`<`TRouteProps`\> |
| `nextState` | `Readonly`<`TRouteState`\> |
| `nextContext` | `any` |

#### Returns

`boolean`

#### Inherited from

React.Component.shouldComponentUpdate

#### Defined in

node_modules/@types/react/index.d.ts:635
