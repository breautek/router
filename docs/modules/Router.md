[@breautek/router - v2.0.1](../README.md) / [Exports](../modules.md) / Router

# Module: Router

## Table of contents

### Classes

- [Router](../classes/Router.Router-1.md)

### Interfaces

- [IRouterProps](../interfaces/Router.IRouterProps.md)
- [IRouterState](../interfaces/Router.IRouterState.md)

### Functions

- [getRouter](Router.md#getrouter)

## Functions

### getRouter

â–¸ `Let` **getRouter**(): [`RouterStrategy`](../classes/RouterStrategy.RouterStrategy-1.md)

#### Returns

[`RouterStrategy`](../classes/RouterStrategy.RouterStrategy-1.md)

#### Defined in

[src/Router.tsx:279](https://github.com/breautek/router/blob/f6dfe6e/src/Router.tsx#L279)
