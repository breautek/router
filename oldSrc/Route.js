'use strict';

import {Component} from 'react';

class Route extends Component {
    constructor(props) {
        super(props);
        this._node = null;
    }

    getNode() {
        var node = this._node;

        while (node && node instanceof Route) {
            node = node.getNode();
        }

        return node;
    }

    render() {
        return this.getComponentsToRender(this);
    }

    /**
     * 
     * @param {React.Component} component 
     */
    getComponentsToRender(component) {
        var url = component.props.url;
        var base = component.props.base || '';

        var ViewComponent = component.props.component;
        var child;

        // console.log('URL', url);
        var routeComponent = component.props.matcher.match(url, this._getChildren(component), base);
        if (routeComponent) {
            child = this.getComponentsToRender(routeComponent);
        }

        return (
            <ViewComponent {...component.props.componentProps} ref={(node) => {
                this._node = node;
            }}>
                {child}
            </ViewComponent>
        );
    }

    /**
     * 
     * @param {React.Component} component 
     */
    _getChildren(component) {
        var children = null;

        if (!component) {
            component = this;
        }
        
        if (!component.props.children) {
            return [];
        }

        if (component.props.children instanceof Array) {
            children = component.props.children;
        }
        else {
            children = [component.props.children];
        }

        return children;
    }
}

export default Route;
export {
    Route
}
