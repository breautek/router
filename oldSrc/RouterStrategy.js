'use strict';

import {EventEmitter} from 'events';

export const EVENT_URL_CHANGE = 'urlchange';

class RouterStrategy extends EventEmitter {

    constructor(router) {
        super();
        this._router = router;
    }

    getRouter() {
        return this._router;
    }

    /**
     * 
     * @param {string} url 
     */
    setTitle(title) {
        var head = document.head.getElementsByTagName('title')[0];

        if (!head) {
            head = document.createElement('title');
            document.head.appendChild(head);
        }

        if (title && title.toString()) {
            head.innerText = title.toString();
        }
    }

    /**
     * 
     * @param {Function} url 
     */
    addURLChangeCallback(callback) {
        this.on(EVENT_URL_CHANGE, callback);
    } 

    /**
     * 
     * @param {Function} url 
     */
    removeURLChangeCallback(callback) {
        this.removeListener(EVENT_URL_CHANGE, callback);
    }

    /**
     * 
     * @param {number} url 
     */
    getLocationAt(position) {
        throw new Error('getLocationAt is abstract.');
    }

    getLocation() {
        throw new Error('getLocation is abstract.');
    }

    getHistoryLength() {
        throw new Error('getHistoryLength is abstract.');
    }

    getScrollRestoration() {
        throw new Error('getScrollRestoration is abstract.');
    }

    go(to) {
        throw new Error('go is abstract.');
    }

    canGo(to) {
        throw new Error('canGo is abstract.');
    }

    canBack() {
        throw new Error('canBack is abstract.');
    }

    canForward() {
        throw new Error('canForward is abstract.');
    }

    peek(to) {
        throw new Error('peek is abstract.');   
    }

    peekForward() {
        throw new Error('peekForward is abstract.');
    }

    peekBack() {
        throw new Error('peekBack is abstract.');
    }

    /**
     * 
     * @param {string} url
     * @param {Object} state 
     */
    pushState(url, state) {
        throw new Error('pushState is abstract.');
    }

    /**
     * 
     * @param {string} url
     * @param {Object} state 
     */
    replaceState(url, state) {
        throw new Error('replaceState is abstract.');
    }

    clear() {
        throw new Error('clear is abstract.');
    }

    /**
     * 
     * @param {string} url 
     */
    _fireURLChange(url) {
        this.emit(EVENT_URL_CHANGE, url);
    }
}

export {RouterStrategy};
export default RouterStrategy;
