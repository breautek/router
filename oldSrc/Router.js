'use strict';

import {Component, cloneElement} from 'react';
import RouterStrategy from './RouterStrategy';
import DefaultStrategy from './DefaultStrategy';
import RouteMatcher from './RouteMatcher';

var instance = null;

export var getRouter = () => {
    if (!instance) {
        return null;
    }

    return instance.getRouterStrategy();
};

class Router extends Component {
    constructor(props) {
        super(props);

        var Strategy;
        if (!props.strategy) {
            Strategy = DefaultStrategy;
        }
        else {
            Strategy = props.strategy;
        }

        var strategy = new Strategy(this);

        this.state = {
            strategy: strategy,
            url : strategy.getLocation()
        };

        this._lastRenderedRoute = null;

        this._onURLChange = this._onURLChange.bind(this);
        this.matcher = new RouteMatcher(strategy);
    }

    getRouterStrategy() {
        return this.state.strategy;
    }
    
    _onURLChange(url) {
        if (url !== this.state.url) {
            this.setState({
                url: url,
                shouldTransition : true
            });
        }
    }

    UNSAFE_componentWillMount() {
        instance = this;
        this.state.strategy.addURLChangeCallback(this._onURLChange);
    }

    componentWillUnmount() {
        instance = null;
        this.state.strategy.removeURLChangeCallback(this._onURLChange);
    };

    UNSAFE_componentWillReceiveProps(nextProps) {
        if (nextProps.strategy && (this.state.strategy instanceof nextProps.strategy)) {
            this.state.strategy.removeURLChangeCallback(this._onURLChange);
            var strat = new nextProps.strategy(this);
            strat.addURLChangeCallback(this._onURLChange);
            this.setState({
                strategy: strat
            });
        }
    }

    render() {
        var currentRoute = this.matcher.match(this.state.url || '/', this._getChildren(), '', this._getIndexRoute());
        var Root = null;
        if (this.props.component) {
            Root = this.props.component;
        }

        if (
            this.state.shouldTransition && (
                currentRoute.props.entryTransition || (
                    this._lastRenderedRoute && 
                    this._lastRenderedRoute.props.exitTransition
                )
            )
        ) {
            this._awaitingTransition = true;
            var exiting = null;
            if (this._lastRenderedRoute) {
                exiting = cloneElement(this._lastRenderedRoute, {
                    ref : (node) => {
                        this._exitingNode = node;
                    }
                });
            }

            // Incoming will always be safe to render, hence no defensive checks
            var incoming = cloneElement(currentRoute, {
                ref : (node) => {
                    this._incomingNode = node;
                }
            });

            if (Root) {
                return <Root router={this.getRouterStrategy()} url={this.state.url}>{[exiting, incoming]}</Root>;
            }
            else {
                return [exiting, incoming];
            }
        }
        else {
            this._lastRenderedRoute = currentRoute;

            if (Root) {
                return <Root router={this.getRouterStrategy()} url={this.state.url}>{currentRoute}</Root>;
            }
            else {
                return currentRoute;
            }
        }
    }

    componentDidUpdate() {
        if (this._awaitingTransition) {
            this._awaitingTransition = false;
            let exitTransitionPromise = null;
            let entryTransitionPromise = null;
            if (this._exitingNode && this._exitingNode.props.exitTransition) {
                exitTransitionPromise = this._exitingNode.props.exitTransition.execute(this._incomingNode.getNode(), this._exitingNode.getNode());
            }
            else {
                exitTransitionPromise = Promise.resolve();
            }

            if (this._incomingNode.props.entryTransition) {
                entryTransitionPromise = this._incomingNode.props.entryTransition.execute(this._incomingNode.getNode(), this._exitingNode.getNode());
            }
            else {
                entryTransitionPromise = Promise.resolve();
            }

            exitTransitionPromise.then(() => {
                return entryTransitionPromise;
            }).catch((error) => {
                console.error(error);
            }).then(() => {
                this._incomingNode = null;
                this._exitingNode = null;
                this.setState({shouldTransition: false});
            });
        }
    }

    getHistoryLength() {
        return this.state.strategy.getHistoryLength();
    }

    getScrollRestoration() {
        return this.state.strategy.getScrollRestoration();
    }

    go(to) {
        this.state.strategy.go(to);
    }

    back() {
        this.state.strategy.go(-1);
    }

    forward() {
        this.state.strategy.go(1);
    }

    _getChildren() {
        var children = null;

        if (this.props.children instanceof Array) {
            children = this.props.children;
        }
        else {
            children = [this.props.children];
        }

        return children;
    }

    _getIndexRoute() {
        var children = this._getChildren();
        for (var i = 0; i < children.length; i++) {
            var child = children[i];
            if (child.props.index) {
                return child;
            }
        }

        return null;
    }
}

export {
    Router
};
export default Router;
