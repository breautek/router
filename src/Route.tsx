
import * as React from 'react';
import { View } from './View';
import { RouteMatcher } from './RouteMatcher';
import { RouterStrategy } from './RouterStrategy';
import { TransitionStrategy } from 'TransitionStrategy';

export interface IRouteProps<T> {
    url: string;
    component: React.ComponentClass<any>;
    index?: boolean;
    entryTransition?: TransitionStrategy;
    exitTransition?: TransitionStrategy;

    base?: string;
    componentProps?: T;
    matcher?: RouteMatcher;
}

export interface IComponentProps {
    url: string;
    router: RouterStrategy;

    //Generally these keys come from URl parameters, and therefore they will
    //be strings, however typescript does not allow us to define this as strings,
    //and have other defined non-string keys.
    [key: string]: any;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface IRouteState {

}

/**
 * This class represents a route that renders a {@link View} component
 */
export class Route<TComponentProps extends IComponentProps = IComponentProps, TRouteProps extends IRouteProps<TComponentProps> = IRouteProps<TComponentProps>, TRouteState extends IRouteState = IRouteState> extends React.Component<TRouteProps, TRouteState> {
    private _node: View;

    constructor(props: TRouteProps) {
        super(props);
        this.render
    }

    public render(): React.ReactNode {
        return this._getComponentsToRender(this);
    }

    public getView(): View {
        return this._node;
    }

    private _getComponentsToRender(component: React.ReactElement | React.Component): React.ReactNode {
        let url: string = component.props.url;
        let base: string = component.props.base || '';

        // eslint-disable-next-line @typescript-eslint/naming-convention
        let ViewComponent: React.ElementType = component.props.component;
        let child: React.ReactNode;

        let routeComponent = component.props.matcher.match(url, this._getChildren(component), base);
        if (routeComponent) {
            child = this._getComponentsToRender(routeComponent);
        }

        return (
            <ViewComponent
                {...component.props.componentProps}
                ref={(node: React.Component) => {
                    if (node) {
                        if (node instanceof View) {
                            this._node = node;
                        }
                        else {
                            throw new Error('Routed components should be a View, but got ' + Object.getPrototypeOf(node).constructor.name + ' instead.');
                        }
                    }
                    else {
                        this._node = null;
                    }
                }}
            >
                {child}
            </ViewComponent>
        );
    }

    private _getChildren(component: React.Component | React.ReactElement): React.ReactElement[] {
        let children: React.ReactElement[] = null;

        if (!component) {
            component = this;
        }
        
        if (!component.props.children) {
            return [];
        }

        if (component.props.children instanceof Array) {
            children = component.props.children as React.ReactElement[];
        }
        else {
            children = [ component.props.children as React.ReactElement ];
        }

        return children;
    }
}
