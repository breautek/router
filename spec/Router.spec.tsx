
import * as React from 'react';
import renderer, {ReactTestRenderer} from 'react-test-renderer';
import {
    TestApp,
    View1,
    View2,
    View3,
    View4,
    OuterView,
    InnerView
} from './env';

import {Router, IRouterProps} from '../src/Router';
import {Route} from '../src/Route';
import { RouterStrategy } from '../src/RouterStrategy';

const tick = function(fn: () => void): void {
    setTimeout(fn, 1);
};

const getTitle = function(): string {
    return document.head.getElementsByTagName('title')[0].innerText;
}

describe('@breautek/router', () => {
    let app: ReactTestRenderer | undefined;

    let props: IRouterProps = {
        component: TestApp
    };

    const router = (): ReactTestRenderer => {
        if (!app) {
            app = renderer.create(
                <Router {...props}>
                    <Route key="page1" url="/page1" component={View1} index />
                    <Route key="page2" url="/page2" component={View2} />
                    <Route key="page3" url="/page3" component={View3} />
                    <Route key="page4" url="/page4" component={View4} />
                </Router>
            );
        }

        return app;
    }

    beforeEach(() => {
        window.location.hash = '/';
        app = undefined;
        props = {
            component: TestApp
        };
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });

    describe('on no routes', () => {
        it('It renders index page', (done) => {
            const comp: ReactTestRenderer = router();
            expect(comp.toJSON()).toMatchSnapshot();
            tick(() => {
                expect(getTitle()).toBe('View1');
                comp.unmount();
                done();
            });
        });

        it('renders index through onNoRoute hook', (done) => {
            props.onNoRoute = (index: React.ReactElement, routes: React.ReactElement[]): React.ReactElement => {
                return index;
            };
            jest.spyOn(props, 'onNoRoute');

            const comp: ReactTestRenderer = router();
            expect(comp.toJSON()).toMatchSnapshot();
            tick(() => {
                expect(props.onNoRoute).toHaveBeenCalled();
                expect(getTitle()).toBe('View1');
                comp.unmount();
                done();
            });
        });

        it('renders null for index through onNoRoute hook', (done) => {
            props.onNoRoute = (index: React.ReactElement, routes: React.ReactElement[]): React.ReactElement | null => {
                return null;
            };
            jest.spyOn(props, 'onNoRoute');
            jest.spyOn(console, 'warn').mockImplementation(() => {});

            const comp: ReactTestRenderer = router();
            expect(comp.toJSON()).toMatchSnapshot();
            tick(() => {
                expect(props.onNoRoute).toHaveBeenCalled();
                expect(console.warn).toHaveBeenCalledWith('No routes matched, and no index route available.');
                comp.unmount();
                done();
            });
        });

        it('renders a different view for index through onNoRoute hook', (done) => {
            props.onNoRoute = (index: React.ReactElement, routes: React.ReactElement[]): React.ReactElement => {
                return routes[2]
            };
            jest.spyOn(props, 'onNoRoute');

            const comp: ReactTestRenderer = router();
            expect(comp.toJSON()).toMatchSnapshot();
            tick(() => {
                expect(props.onNoRoute).toHaveBeenCalled();
                comp.unmount();
                done();
            });
        });
    });

    it('getInstance()', () => {
        expect(Router.getInstance()).toBe(null);
        const comp: ReactTestRenderer = router();

        // Ideally we want to ensure the internal state matches what is returned
        // in the public API, but Facebook testing mantra is to test props and
        // their behaviour of those props rather than concerning itself with
        // implementation details, which state is considered an implementation
        // detail of a component.
        // https://stackoverflow.com/a/61813414/4685664
        expect(Router.getInstance()).toBeInstanceOf(RouterStrategy);
        comp.unmount();
    });

    it('can navigate pages', (done) => {
        const comp: ReactTestRenderer = router();

        const r: RouterStrategy = Router.getInstance();
        expect(r.getHistoryLength()).toBe(0);
        expect(r.canBack()).toBe(false);

        const urlChange = (url: string) => {
            r.removeURLChangeCallback(urlChange);

            tick(() => {
                expect(getTitle()).toBe('View2');
                expect(url).toBe('/page2');
                expect(r.canBack()).toBe(true);
                expect(r.getHistoryLength()).toBe(2);
                expect(comp.toJSON()).toMatchSnapshot();

                comp.unmount();
                done();
            });
        };
        
        r.addURLChangeCallback(urlChange);
        tick(() => {
            r.pushState('/page2');
        });
    });

    it('URLChangeCallback should fire once per URL change', (done) => {
        const comp: ReactTestRenderer = router();
        const r: RouterStrategy = Router.getInstance();
        const onUrlChange = jest.fn();
        
        tick(() => {
            r.addURLChangeCallback(onUrlChange);
            r.pushState('/page');
            tick(() => {
                expect(onUrlChange).toHaveBeenCalledTimes(1);
                expect(onUrlChange).toHaveBeenCalledWith<[string]>('/page');
                comp.unmount();
                r.removeURLChangeCallback(onUrlChange);
                done();
            });
        });
    });

    describe('subviews', () => {
        it('can render subviews', (done) => {
            app = renderer.create(
                <Router component={TestApp}>
                    <Route key="outer" url="/outerView" component={OuterView} index>
                        <Route key="inner" url="/innerView" component={InnerView} />
                    </Route>
                </Router>
            );
    
            const r: RouterStrategy = Router.getInstance();

            tick(() => {
                r.pushState('/outerView/innerView');
                tick(() => {
                    expect(app.toJSON()).toMatchSnapshot();
                    app.unmount();
                    done();
                });
            });
        });
    });
});
