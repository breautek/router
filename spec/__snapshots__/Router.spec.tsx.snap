// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@breautek/router can navigate pages 1`] = `
<div
  className="app"
>
  <div
    className="View"
    style={{}}
  >
    View2
  </div>
</div>
`;

exports[`@breautek/router on no routes It renders index page 1`] = `
<div
  className="app"
>
  <div
    className="View View1"
    style={{}}
  >
    View1
  </div>
</div>
`;

exports[`@breautek/router on no routes renders a different view for index through onNoRoute hook 1`] = `
<div
  className="app"
>
  <div
    className="View"
    style={{}}
  >
    View3
  </div>
</div>
`;

exports[`@breautek/router on no routes renders index through onNoRoute hook 1`] = `
<div
  className="app"
>
  <div
    className="View View1"
    style={{}}
  >
    View1
  </div>
</div>
`;

exports[`@breautek/router on no routes renders null for index through onNoRoute hook 1`] = `
<div
  className="app"
/>
`;

exports[`@breautek/router subviews can render subviews 1`] = `
<div
  className="app"
>
  <div
    className="View"
    style={{}}
  >
    <div>
      Outer View
    </div>
    <div
      className="View"
      style={{}}
    >
      Inner View
    </div>
  </div>
</div>
`;
